*Практические* задания к уроку №6

###Тема “Обучение с учителем”







Задание 1.

Импортируйте библиотеки pandas и numpy.

Загрузите "Boston House Prices dataset" из встроенных наборов данных библиотеки sklearn. 
Создайте датафреймы X и y из этих данных.

Разбейте эти датафреймы на тренировочные (X_train, y_train) и тестовые (X_test, y_test) с помощью функции train_test_split так, чтобы размер тестовой выборки
составлял 30% от всех данных, при этом аргумент random_state должен быть равен 42.

Создайте модель линейной регрессии под названием lr с помощью класса LinearRegression из модуля sklearn.linear_model.

Обучите модель на тренировочных данных (используйте все признаки) и сделайте предсказание на тестовых.

Вычислите R2 полученных предказаний с помощью r2_score из модуля sklearn.metrics



In [ ]:
#импортируем необходимые библиотеки
import numpy as np
import pandas as pd


In [ ]:
#загрузим  dataset
from sklearn.datasets import load_boston
boston = load_boston()
data = boston.data

In [ ]:
#boston.keys()
#data.shape

In [ ]:
target = boston.target
feature_names = boston.feature_names

In [ ]:
#создадим датафреймы
X = pd.DataFrame(data, columns=feature_names)
y = pd.DataFrame(target, columns=['price'])

In [ ]:
#разобьём датафреймы на тренировочные и тестовые
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

#создадим модель линейной регрессии
from sklearn.linear_model import LinearRegression
lr = LinearRegression()

In [ ]:
#обучение модели
lr.fit(X_train, y_train)

LinearRegression()

In [ ]:
y_pred = lr.predict(X_test)   #предсказание данных
check_test = pd.DataFrame({'y_test': y_test['price'],
                          'y_pred': y_pred.flatten()},
                          columns=['y_test', 'y_pred'])

check_test.head(10)     #сравним тестовую выборку и предсказанные данные

y_test     y_pred
173    23.6  28.648960
274    32.4  36.495014
491    13.6  15.411193
72     22.8  25.403213
452    16.1  18.855280
76     20.0  23.146689
316    17.8  17.392124
140    14.0  14.078599
471    19.6  23.036927
500    16.8  20.599433

In [ ]:
from sklearn.metrics import r2_score
#Вычислим R2 полученных предказаний
r2_score(y_test, y_pred)

0.7112260057484974

Задание 2.

Создайте модель под названием model с помощью RandomForestRegressor из модуля sklearn.ensemble.

Сделайте агрумент n_estimators равным 1000,
max_depth должен быть равен 12 и random_state сделайте равным 42.

Обучите модель на тренировочных данных аналогично тому, как вы обучали модель LinearRegression,
но при этом в метод fit вместо датафрейма y_train поставьте y_train.values[:, 0,
чтобы получить из датафрейма одномерный массив Numpy,
так как для класса RandomForestRegressor в данном методе для аргумента y предпочтительно применение массивов вместо датафрейма.

Сделайте предсказание на тестовых данных и посчитайте R2. 
Сравните с результатом из предыдущего задания.

Напишите в комментариях к коду, какая модель в данном случае работает лучше.



In [ ]:
#импортируем необходимые библиотеки
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import roc_auc_score

In [ ]:
#создадим модель
model = RandomForestRegressor(n_estimators=1000,
                              max_depth=12,
                              random_state=42)

In [ ]:
#обучение модели
model.fit(X_train, y_train.values[:, 0])

RandomForestRegressor(max_depth=12, n_estimators=1000, random_state=42)

In [ ]:
y_pred = model.predict(X_test)   #предсказание данных
check_test = pd.DataFrame({'y_test': y_test['price'],
                          'y_pred': y_pred.flatten()},
                          columns=['y_test', 'y_pred'])

check_test.head(10)     #сравним тестовую выборку и предсказанные данные

y_test     y_pred
173    23.6  22.806412
274    32.4  31.131464
491    13.6  16.339125
72     22.8  23.810726
452    16.1  17.139521
76     20.0  21.832284
316    17.8  19.895747
140    14.0  14.754118
471    19.6  21.240835
500    16.8  20.898658

In [ ]:
from sklearn.metrics import r2_score
#Вычислим R2 полученных предказаний
r2_score(y_test, y_pred)

0.87472606157312

Первая модель дала результат 0.7112   (LinearRegression)

Вторая модель дала результат 0.8747   ( RandomForestRegressor)


Вывод:

в данному случае более эффективна модель  RandomForestRegressor